add frame skip/frozen frame data. refactor the way charts are created

bbaldino opened the pull request.

bbaldino committed:
add frame skip/frozen frame data. refactor the way charts are created

bbaldino commented on app.js:
```diff
@@ -9,11 +8,71 @@ class Dashboard extends React.Component {
      */
+    constructor(props) {
+        super(props);
+        this.psnrTestResultTransformFunc = jsonData => {
```
this (defining this all here) doesn't feel great to me.  curious if you have any suggestions.

virtuacoplenny replied:
Why are you defining it in the constructor?

bbaldino replied:
i tried as a member but was running into problems for some reason, probably just syntax mistakes on my part.  i don't think it feels great as a member either though, really.

virtuacoplenny replied:
To be fair, I think previously you were mixing view logic with model logic.

virtuacoplenny commented on app.js:
```diff
@@ -9,11 +8,71 @@ class Dashboard extends React.Component {
+            const frameSkipData
+                = extractPctData(jsonData, 'numSkippedFrames', 'totalFrames');
+            const frameFrozenData
+                = extractPctData(jsonData, 'numFrozenFrames', 'totalFrames');
```
This almost sounds like it needs a chart model to abstract away this info.

bbaldino replied:
you mean a chart model like...the ones i had that you didn't like? :)

but yeah, all this here doesn't feel good and feels like it should be somewhere more specific

virtuacoplenny replied:
To be fair, I think previously you were mixing view logic with model logic.

virtuacoplenny commented on charts/baseChart.js:
```diff
@@ -28,9 +29,26 @@ module.exports = class Chart extends React.Component {
+                        text: this.props.config.graphXAxis
+                    }
+                },
+                series: this.state.data
```
What do you think about moving these values off of config and onto props instead? props is essentially a config already.

virtuacoplenny commented on charts/baseChart.js:
```diff
@@ -28,9 +29,26 @@ module.exports = class Chart extends React.Component {
+                },
+                xAxis: {
+                    title: {
+                        text: this.props.config.graphXAxis
```
What do you think about moving these values off of config and onto props instead? props is essentially a config already.

bbaldino replied:
actually i'm realizing now i've got a bit of a problem (mixing different y value scales).  i didn't really want to do it as 2 separate charts but might have to.  i'll look into that.

virtuacoplenny commented on charts/psnrChart.js:
```diff
@@ -1,41 +0,0 @@
-            .then(jsonResult => this.setState({ data: jsonResult }))
-            .catch(err => this.setState({ error: err }));
-    }
-};
```
You were right. I do like this change.

virtuacoplenny commented on charts/baseChart.js:
```diff
@@ -40,17 +58,34 @@ module.exports = class Chart extends React.Component {
+
+        this._retrieveResultsHelper(config.resultsUrl, config.resultsFunc)
+            .then(jsonResult => this.setState({ data: jsonResult }))
+            .catch(err => this.setState({ error: err }));
```
just a warning, and one that is for thought and not action at this point, this is mixing data handling and presentation logic. Normally the two should be separated, if convenient, to avoid having the smarts of the application "hidden" within what should be simpler presentation logic. Ideally presentation logic just takes in data and displays it in some manner.

bbaldino replied:
if there's a better way to trigger the data retrieval let me know. this was based on an example i'd found.

virtuacoplenny commented on routes/psnr.js:
```diff
@@ -2,16 +2,29 @@ const PsnrResult = require('../models/psnrResult');
-            psnrValue
+            psnr,
+            numFrozenFrames,
+            numSkippedFrames,
```
Are route/frameSkip and models/frameSkipResult still used?

bbaldino replied:
nope, forgot to get rid of those.

virtuacoplenny commented on app.js:
```diff
@@ -9,11 +8,71 @@ class Dashboard extends React.Component {
+                json.reduce((currData, currRow) => {
+                    const field = currRow[yAxisFieldName];
+
+                    currData.push([ currRow.buildNum, field ]);
```
Is it a problem that you're modifying the passed array in place? I don't see it being a problem...but I want to be sure because that's caused problems for me in other codebases.

bbaldino replied:
isn't this always the case with `reduce` ?

virtuacoplenny replied:
I'm referring specifically to the currData.push.

virtuacoplenny commented on app.js:
```diff
@@ -9,11 +8,71 @@ class Dashboard extends React.Component {
+                }
+            ];
+        };
+    }
```
This is pretty clever. At first I was kind of worried because maybe the relationship between this parser function and the chart is separated too much, but the more I read it the more I think it works well enough.

bbaldino committed:
pr feedback, some refactoring

bbaldino committed:
add a comment, remove unused files

virtuacoplenny commented on app.js:
```diff
@@ -12,8 +13,34 @@ class Dashboard extends React.Component {
+                    config={{
+                        resultsUrl: './psnrResults',
+                        resultTransformers: [ psnrTransformer ]
+                    }}
```
Do you still prefer having props.config over having props.resultsUrl and props.resultsTransformers?

bbaldino replied:
i'd liked this split better because the stuff here in props are final values, while what's in config isn't used as part of the actual chart itself, they're just means to an end...so this division felt cleaner to me.

virtuacoplenny commented on app.js:
```diff
@@ -12,8 +13,34 @@ class Dashboard extends React.Component {
+                    graphYAxis='% of total frames'
+                    graphXAxis='Build number'
+                    config={{
+                        resultsUrl: './psnrResults',
```
This is an interesting tidbit. Two requests are made to the same endpoint?

bbaldino replied:
yeah, this felt a bit weird but not doing it this way means we'd have to do something like cache it and have a layer which would transparently check the cache.  if you've got a good idea let me know.  i don't think it's a dealbreaker for this, but it would be nice to be able to avoid it.

virtuacoplenny replied:
Creating a model that does the request and parses out various results. The model doesn't have to be a conventional model in the sense of mvc. Dashboard (or whatever other view) calls fetch through some api on the model, which returns a promise, then you can set state to update the chart display.

That way you at least separate out more of your business logic from your view logic.

virtuacoplenny commented on charts/psnrChartHandler.js:
```diff
@@ -0,0 +1,24 @@
+ * @returns an object suitable to be placed in a series array
+ * for a highchart
+ */
+export default json => {
```
Why not break the fetch out as well, maybe into this helper?

bbaldino replied:
hmm but then the fetch logic would be duplicated for each chart

bbaldino committed:
refactor things a bit

this gets rid of the duplicate request for the psnr test result data by
having the dashboard use a result requester to fetch the data once, and
then setting the state for each chart by retrieving the data it needs
from the requester

virtuacoplenny commented on models/psnrResultRequester.js:
```diff
@@ -0,0 +1,99 @@
+module.exports = class PsnrResultRequester {
```
This is the confusing part... This file is actually used by the react app but you're using module.exports which is being used primarily by the node-side files. I'm assuming it all still works because of webpack.
